src/ln/functional_tests.rs

```diff
@@ -5825,6 +5697,240 @@ fn test_dynamic_spendable_outputs_local_htlc_success_tx() {
+					a_done = true;
+					&nodes[0]
+				} else {
+					assert_eq!(*node_id, nodes[1].node.get_our_node_id());
+					assert_eq!(updates.update_fail_htlcs.len(), if announce_latest { 4 } else { 3 });
+					&nodes[1]
+				};
+				target.node.handle_update_fail_htlc(&nodes[2].node.get_our_node_id(), &updates.update_fail_htlcs[0]).unwrap();
+				target.node.handle_update_fail_htlc(&nodes[2].node.get_our_node_id(), &updates.update_fail_htlcs[1]).unwrap();
+				target.node.handle_update_fail_htlc(&nodes[2].node.get_our_node_id(), &updates.update_fail_htlcs[2]).unwrap();
+				if announce_latest {
+					target.node.handle_update_fail_htlc(&nodes[2].node.get_our_node_id(), &updates.update_fail_htlcs[3]).unwrap();
+					if *node_id == nodes[0].node.get_our_node_id() {
+						target.node.handle_update_fail_htlc(&nodes[2].node.get_our_node_id(), &updates.update_fail_htlcs[4]).unwrap();
+					}
+				}
+				commitment_signed_dance!(target, nodes[2], updates.commitment_signed, false, true);
+			},
+			_ => panic!("Unexpected event"),
+		}
+	}
+
+	let as_events = nodes[0].node.get_and_clear_pending_events();
+	assert_eq!(as_events.len(), if announce_latest { 5 } else { 3 });
+	let mut as_failds = HashSet::new();
+	for event in as_events.iter() {
+		if let &Event::PaymentFailed { ref payment_hash, ref rejected_by_dest, .. } = event {
+			assert!(as_failds.insert(*payment_hash));
+			if *payment_hash != payment_hash_2 {
+				assert_eq!(*rejected_by_dest, deliver_last_raa);
+			} else {
+				assert!(!rejected_by_dest);
+			}
+		} else { panic!("Unexpected event"); }
+	}
+	assert!(as_failds.contains(&payment_hash_1));
+	assert!(as_failds.contains(&payment_hash_2));
+	if announce_latest {
+		assert!(as_failds.contains(&payment_hash_3));
+		assert!(as_failds.contains(&payment_hash_5));
+	}
+	assert!(as_failds.contains(&payment_hash_6));
+
+	let bs_events = nodes[1].node.get_and_clear_pending_events();
+	assert_eq!(bs_events.len(), if announce_latest { 4 } else { 3 });
+	let mut bs_failds = HashSet::new();
+	for event in bs_events.iter() {
+		if let &Event::PaymentFailed { ref payment_hash, ref rejected_by_dest, .. } = event {
+			assert!(bs_failds.insert(*payment_hash));
+			if *payment_hash != payment_hash_1 && *payment_hash != payment_hash_5 {
+				assert_eq!(*rejected_by_dest, deliver_last_raa);
+			} else {
+				assert!(!rejected_by_dest);
+			}
+		} else { panic!("Unexpected event"); }
+	}
+	assert!(bs_failds.contains(&payment_hash_1));
+	assert!(bs_failds.contains(&payment_hash_2));
+	if announce_latest {
+		assert!(bs_failds.contains(&payment_hash_4));
+	}
+	assert!(bs_failds.contains(&payment_hash_5));
+
+	let as_msg_events = nodes[0].node.get_and_clear_pending_msg_events();
+	assert_eq!(as_msg_events.len(), if deliver_last_raa { 1 } else if !announce_latest { 3 } else { 5 });
+	let bs_msg_events = nodes[1].node.get_and_clear_pending_msg_events();
+	assert_eq!(bs_msg_events.len(), if deliver_last_raa { 2 } else if !announce_latest { 3 } else { 4 });
```

Comment:
What is the rationale there to get 2 PaymentFailureNetworkUpdate for node B ? Doesn't seem to match with above-dust HTLCs origins ?

Reply:
Added a comment to explain the value.